src/rayon.rs

```diff
@@ -0,0 +1,209 @@
+                .drive(consumer)
+        }
+
+        fn len(&mut self) -> usize {
+            self.entries.len()
+        }
+
+        fn with_producer<CB>(self, callback: CB) -> CB::Output
+            where CB: ProducerCallback<Self::Item>
+        {
+            self.entries.into_par_iter()
+                .map($map_elt)
+                .with_producer(callback)
+        }
+    }
+}
+
+
+impl<K, V, S> IntoParallelIterator for OrderMap<K, V, S>
+    where K: Hash + Eq + Send,
+          V: Send,
+          S: BuildHasher,
+{
+    type Item = (K, V);
+    type Iter = IntoParIter<K, V>;
+
+    fn into_par_iter(self) -> Self::Iter {
+        IntoParIter {
+            entries: self.entries,
+        }
+    }
+}
+
+pub struct IntoParIter<K, V> {
+    entries: Vec<Bucket<K, V>>,
+}
+
+impl<K: Send, V: Send> ParallelIterator for IntoParIter<K, V> {
+    type Item = (K, V);
+
+    parallel_iterator_methods!(|e| (e.key, e.value));
+}
+
+impl<K: Send, V: Send> IndexedParallelIterator for IntoParIter<K, V> {
+    indexed_parallel_iterator_methods!(|e| (e.key, e.value));
+}
+
+
+impl<'a, K, V, S> IntoParallelIterator for &'a OrderMap<K, V, S>
+    where K: Hash + Eq + Sync,
+          V: Sync,
+          S: BuildHasher,
+{
+    type Item = (&'a K, &'a V);
+    type Iter = ParIter<'a, K, V>;
+
+    fn into_par_iter(self) -> Self::Iter {
+        ParIter {
+            entries: &self.entries,
+        }
+    }
+}
+
+pub struct ParIter<'a, K: 'a, V: 'a> {
+    entries: &'a [Bucket<K, V>],
+}
+
+impl<'a, K: Sync, V: Sync> ParallelIterator for ParIter<'a, K, V> {
+    type Item = (&'a K, &'a V);
+
+    parallel_iterator_methods!(|e| (&e.key, &e.value));
+}
+
+impl<'a, K: Sync, V: Sync> IndexedParallelIterator for ParIter<'a, K, V> {
+    indexed_parallel_iterator_methods!(|e| (&e.key, &e.value));
+}
+
+
+impl<'a, K, V, S> IntoParallelIterator for &'a mut OrderMap<K, V, S>
+    where K: Hash + Eq + Sync + Send,
+          V: Send,
+          S: BuildHasher,
+{
+    type Item = (&'a K, &'a mut V);
+    type Iter = ParIterMut<'a, K, V>;
+
+    fn into_par_iter(self) -> Self::Iter {
+        ParIterMut {
+            entries: &mut self.entries,
+        }
+    }
+}
+
+pub struct ParIterMut<'a, K: 'a, V: 'a> {
+    entries: &'a mut [Bucket<K, V>],
+}
+
+impl<'a, K: Sync + Send, V: Send> ParallelIterator for ParIterMut<'a, K, V> {
+    type Item = (&'a K, &'a mut V);
+
+    parallel_iterator_methods!(|e| (&e.key, &mut e.value));
+}
+
+impl<'a, K: Sync + Send, V: Send> IndexedParallelIterator for ParIterMut<'a, K, V> {
+    indexed_parallel_iterator_methods!(|e| (&e.key, &mut e.value));
+}
+
+
+// This form of intermediate collection is also how Rayon collects `HashMap`.
+// Note that the order will also be preserved!
+fn collect<I: IntoParallelIterator>(iter: I) -> LinkedList<Vec<I::Item>> {
+    iter.into_par_iter()
+        .fold(Vec::new, |mut vec, elem| {
+            vec.push(elem);
+            vec
+        })
+        .collect()
+}
```

Comment:
Loll wut!! LinkedList is used in a real-life PR?!! 🥇 🐫 🎈 :shipit: 

(Ahem I'll be back to my regular @bluss in a bit.)

Reply:
You know, it's funny, for all the strategies we benchmarked for Rayon collect, I don't think we actually tried a `Vec<Vec<T>>` intermediate.  I just assumed that a list was best for this purpose because of its O(1) append.  I'm going to have to do some testing now...

Reply:
I've kept the `LinkedList`, but now collect it with map-reduce -- see rayon-rs/rayon#479.